A58-client-side-weighted-round-robin-lb-policy.md

```diff
@@ -0,0 +1,251 @@
+
+### LB Policy Config and Parameters
+
+Add the policy config with parameters as follows. This policy is enabled as the endpoint picking policy of `WrrLocality` in [gRFC A52], see [below](#xds-integration) for more.
+```textproto
+message LoadBalancingConfig {
+   // …
+   WeightedRoundRobin weighted_round_robin_experimental = 18 [json_name = "weighted_round_robin_experimental "];
+}
+
+message WeightedRoundRobin {
+   google.protobuf.Duration blackout_period = 1;
+   google.protobuf.Duration weight_expiration_period = 2;
+   bool enable_oob_load_report = 3;
+   google.protobuf.Duration oob_reporting_period = 4;
+   google.protobuf.Duration weight_update_period = 5;
+}
+
+```
+
+`blackout_period` is to handle backend churns. `weighted_round_robin` rejects load reports until a backend reports valid weights for at least `blackout_period` during which the average weight value is used instead. Disables blackout when set to a negative value. Defaults to 10 seconds.
+
+Load metrics for a subchannel expire after `weight_expiration_period` is passed without any update from the backend, including with per-call or OOB load reporting. Once expired, the average weight value is used instead. Defaults to 3 minutes. Cannot be disabled but the period has no max value allowed.
+
+The client initiates OOB load reporting by sending a request only when `enable_oob_load_report` is set. Defaults to false. When set, this policy does not parse per-call load reports. See below for more on load reporting. Note this does not affect the per-call load reports in xDS LRS.
+
+`oob_reporting_period` is only used when OOB load reporting is used. Note the effective reporting period is determined as outlined in [gRFC A51] and may be different. The client requests the minimum of values from all interested LB policies, but the server enforces its own minimum reporting period. Defaults to 10 seconds.
+
+The client updates effective weights periodically every `weight_update_period`. This parameter defaults to 1 second. The lowest value allowed is 100ms. When set to a value lower than 100ms the period is set to 100ms. Weight updates triggered by changes in the connectivity state or backend addresses are not affected.
+
+### Earliest Deadline First (EDF) Scheduler
+
+`weighted_round_robin` introduces a scheduler that manages weighted subchannel selection. A new scheduler instance is created every time weights are updated.  A state object is passed when created so that the scheduler can start where it last left across multiple instances. The scheduler must be light and cheap to create and allow concurrent pick calls. The scheduler returns the index in the input weight array and the caller has to map the index to a subchannel.
```

Comment:
Done, except I said "atomically updated" instead of saying an atomic. So far I avoided mentioning anything specific to the "static stride" as it is one way of implementing the EDF scheduling, and the state being one atomic is specific to the static stride. I guess we can state that i) the state is persisted and ii) updated after each pick calls (atomically). LMK.